### DISCUSSION TITLE
Lesson 11: Unable to import & view BasicNFT on Metamask

### DISCUSSION THREAD

**Discussion by itznishant** [Cyfrin/foundry-full-course-cu#580] [Q&A]
Lesson 11: Unable to import & view BasicNFT on Metamask

I successfully deployed the Lesson 11 BasicNFT (renamed to MyNFTToken) on Sepolia and minted NFT.
When I try to import this in metamask, nothing happens and no NFT gets added.

I tried clearing the account nonce and activity data but the issue persists.

The contract displays the token_uri for tokenId '0' correctly as well [ref image]:
<img width="776" alt="image" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/17523443/f8a071b4-343f-4b17-ba85-bfcf8337b493">

Details:

The metadata (URI) is uploaded to IPFS :
```ipfs://QmQAeXueHxb786TvEeQiMcfu5sb7ZVr3fFjJZDJQ11ga5M?filename=0-CUTE-PUG.json```

Image link:
```https://ipfs.io/ipfs/QmWyzqfkpfNKQfHqY1ypeZtSTSDBhFDvTbBV1urrMByje6?filename=pug_cute.png```

Contract address:
```https://sepolia.etherscan.io/address/0xdcc9c847b84ca906c670f210c5b8124a1515fa78```

How do I resolve this issue.

**Comment by shikhar229169** [comment]
Actually you need to enter only this as token URI:
```
ipfs://QmQAeXueHxb786TvEeQiMcfu5sb7ZVr3fFjJZDJQ11ga5M
```
We are only required to add the ipfs hash, so remove this part from the token URI: ```?filename=0-CUTE-PUG.json```
And it will show up.

**Comment by itznishant** [comment]
> Actually you need to enter only this as token URI:
> 
> ```
> ipfs://QmQAeXueHxb786TvEeQiMcfu5sb7ZVr3fFjJZDJQ11ga5M
> ```
> 
> We are only required to add the ipfs hash, so remove this part from the token URI: `?filename=0-CUTE-PUG.json` And it will show up.

That's not the issue here. The link with filename loads up absolutely fine, anyone can check this.

<img width="864" alt="img" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/17523443/237135c5-7239-468c-933d-a8e7b0542a09">

Issue seems to be with Metamask itself. There is no option to add tokenuri, you only add contract address and tokenId to import NFT.

Are people here able to add this BasicNFT to other wallets, does that show up correctly?

**Reply by theirrationalone** [reply]
No Others can't add Your NFT Because It's only Owned by You. If We try to import your NFT then, Ownership issue will rise.

**Reply by shikhar229169** [reply]
Actually I tried to mint NFT with both URIs the one that you mentioned and the one in which I removed ```?filename=0-CUTE-PUG.json```.

So, for the token URI that you mentioned I got token Id as 2 and metamask says NFT added but it was not there.
Then I minted another NFT at token id - 3, in which I removed ```?filename=0-CUTE-PUG.json```, and it was added to metamask and also it was visible.

So, consider checking with that.

One more thing I am not saying to add token URI in metamask. I am talking about the function call:
```solidity
function mintNft(string memory _tokenUri) public {
        s_tokenIdToURI[s_tokenCounter] = _tokenUri;

        _safeMint(msg.sender, s_tokenCounter);
        s_tokenCounter++;
}
```

When you are calling the mintNft function, you need to remove ```?filename=0-CUTE-PUG.json``` while passing tokenURI to the argument of this function.
So instead of passing ```ipfs://QmQAeXueHxb786TvEeQiMcfu5sb7ZVr3fFjJZDJQ11ga5M?filename=0-CUTE-PUG.json```, pass this: ```ipfs://QmQAeXueHxb786TvEeQiMcfu5sb7ZVr3fFjJZDJQ11ga5M```

**Reply by itznishant** [reply]
Thanks for minting those tokens and checking this. Highly appreciate it!

Yes, I noticed your point on the filename in the uri. However, I don't see an issue since the course codebase has a similar uri & that NFT got imported fine ? 🤔 

https://github.com/Cyfrin/foundry-nft-f23/blob/fd791c5312fffa20af7a139093e9b7cf2d23cc73/script/Interactions.s.sol#L11C27-L11C27

Also, I checked out a web3 project's NFT contract on mainnet, which has another interesting issue.

It seems there are two **different** ipfs hash generated for the **same json metadata**.

If you upload an entire folder to ipfs consisting of multiple json metadata files, it seems a cid hash is generated for the base directory and then for each file based on the filename.

```https://ipfs.io/ipfs/bafybeiafwvc5wsrloxgwijuxnswpj2iqm4yc2sofy7yppivgxjgbzgepr4/```

<img width="1328" alt="__simg" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/17523443/4579cc35-45ba-4205-9eea-5a27f62c6cb7">

Use the link above to check metadata for both 1.json and 'bafk...72ea' hash next to it. 
They're the same metadata file with 2 different hashes , one with a filename added at the end!

This added a bit of confusion, thanks a lot ipfs!

**Comment by itznishant** [comment]
> No Others can't add Your NFT Because It's only Owned by You. If We try to import your NFT then, Ownership issue will rise.

I meant adding  an NFT minted from BasicNFT contract, which you own !

You'll not be able to import 'my' NFT, no offence.
NFT is a unique token which actually is owned by the private key of the EOA.
BasicNFT is the contract which holds these unique tokens for us.
Both are two different things.

If it was possible to just import anyone's assets on a blockchain, I'd probably not be in the web3 scene lol.
There's a reason we have decentralised ownership of assets on a blockchain, which gives us, the participants, the right to own our own assets.

**Comment by FurkanSezal** [comment]
Did you able to solve this? I basicly code with patrick and copy paste it from repo getting same issue. Nft was minted but i cant display on metamask
This the contract address on sepolia : 0xadDC3544Ec772589eE5DCeB6596Ad0143C7C2a06
Here is script: 

```
// SPDX-License-Identifier: MIT
pragma solidity 0.8.19;

import {Script} from "forge-std/Script.sol";
import {BasicNft} from "../src/BasicNft.sol";
import {console} from "forge-std/console.sol";

import {DevOpsTools} from "lib/foundry-devops/src/DevOpsTools.sol";

contract MintBasicNft is Script {
    string public constant PUG_URI =
        "ipfs://bafybeig37ioir76s7mg5oobetncojcm3c3hxasyd4rvid4jqhy4gkaheg4/?filename=0-PUG.json";
    uint256 deployerKey;

    function run() external {
        address mostRecentlyDeployedBasicNft = DevOpsTools
            .get_most_recent_deployment("BasicNft", block.chainid);
        mintNftOnContract(mostRecentlyDeployedBasicNft);
        console.log(mostRecentlyDeployedBasicNft);
    }

    function mintNftOnContract(address basicNftAddress) public {
        vm.startBroadcast();
        BasicNft(basicNftAddress).mintNft(PUG_URI);
        vm.stopBroadcast();
    }
}

```
